### ISSUE TITLE
Synapse not honoring .well-known/matrix/server for invitations

### ISSUE THREAD

**Issue by ndarilek** [matrix-org/synapse#5292]
Using the latest Synapse on both servers, I just tried inviting myself from one server to another and received this traceback. Looking at the logs, it seems like Synapse is disregarding my .well-known setup at https://thewordnerd.info/.well-known/matrix/server. Not sure if it is due to the redirect, but if Synapse could follow it, I'm guessing the file at the end is valid. Here is the log:

```
2019-05-30 14:41:28,259 - synapse.http.matrixfederationclient - 465 - WARNING - GET-142- {GET-O-1} [thewordnerd.info] Request failed: GET matrix://thewordnerd.info/_matrix/federation/v1/query/profile?user_id=%40nolan%3Athewordnerd.info&field=displayname: ResponseNeverReceived:[CancelledError()]
2019-05-30 14:41:28,263 - synapse.http.server - 112 - ERROR - GET-142- Failed handle request via 'ProfileRestServlet': <SynapseRequest at 0x7fa38bb08080 method='GET' uri='/_matrix/client/r0/profile/%40nolan%3Athewordnerd.info' clientproto='HTTP/1.1' site=8008>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/synapse/http/matrixfederationclient.py", line 404, in _send_request
    response = yield request_deferred
twisted.web._newclient.ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/synapse/http/server.py", line 81, in wrapped_request_handler
    yield h(self, request)
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/http/server.py", line 316, in _async_render
    callback_return = yield callback(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/rest/client/v1/profile.py", line 141, in on_GET
    displayname = yield self.profile_handler.get_displayname(user)
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/handlers/profile.py", line 140, in get_displayname
    ignore_backoff=True,
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/federation/transport/client.py", line 202, in make_query
    ignore_backoff=ignore_backoff,
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/http/matrixfederationclient.py", line 731, in get_json
    timeout=timeout,
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/http/matrixfederationclient.py", line 248, in _send_request_with_optional_trailing_slash
    request, **send_request_args
  File "/usr/local/lib/python3.6/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/usr/local/lib/python3.6/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/usr/local/lib/python3.6/site-packages/synapse/http/matrixfederationclient.py", line 409, in _send_request
    raise_from(RequestSendFailed(e, can_retry=True), e)
  File "<string>", line 3, in raise_from
synapse.api.errors.RequestSendFailed: Failed to send request: ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]
```

**Comment by anoadragon453** [comment]
It could possibly be because the Content-Type header for https://www.thewordnerd.info/.well-known/matrix/server is `text/html` instead of `application/json`. Could you try changing your webserver settings and see if it fixes the issue?

If so, the fix is relatively straight-forward.

**Comment by ndarilek** [comment]
In the interest of being thorough, the content-type was text/plain 
according to my Curl run. :) I set it to application/json, same result.


Is it possible that Synapse doesn't follow 301/302 redirects? That would 
be unfortunate for those of us redirecting example.com to www.example.com.

**Comment by anoadragon453** [comment]
It definitely *should* be, but I can look into setting up a test page and seeing if it fails.

Can you look in the logs to see if it tried to access `https://thewordnerd.info/.well-known/matrix/server` and if it failed in any way?

**Comment by ndarilek** [comment]
There are no helpful logs above the ones I posted. However, I'm almost 
entirely certain Synapse isn't honoring 301s.


I'm running a Matrix server at https://matrix.lightsout.games. I 
initially had all of https://lightsout.games redirecting to 
https://www.lightsout.games, including /.well-known/matrix.


Initially, https://www.lightsout.games/.well-known/matrix/server served 
as content-type: text/plain. I fixed that, but it still didn't make a 
difference. I couldn't even join rooms on lightsout.games, always 
getting M_UNKNOWN, along with errors that the room was unavailable. The 
only way I could join rooms was by creating a special case to serve 
https://lightsout.games/.well-known/matrix/server while redirecting 
everything else to https://www.lightsout.games.


So, in short, I'm pretty sure whatever handles HTTP requests isn't 
explicitly following 301/302 redirects. I could be wrong, but once that 
change was made, I was able to join rooms. I didn't test invites, but 
given that I couldn't join until I made that change, that seems most 
likely. I realize these are likely separate issues in both Riot and 
Synapse, so testing around 301/302 redirects might be useful on both fronts.

**Comment by richvdh** [comment]
Synapse does follow 30x redirects for `/.well-known/matrix/server`. The tests for it are here: https://github.com/matrix-org/synapse/blob/v1.0.0/tests/http/federation/test_matrix_federation_agent.py#L547.

Please can you:

* make sure you have logging set to at least INFO
* restart synapse to flush the .well-known cache
* perform some action to initiate a new federation request
* send us all the logs corresponding to the failed request (see https://github.com/matrix-org/synapse/wiki#how-can-i-find-the-lines-corresponding-to-a-given-http-request-in-my-homeserver-log)

**Comment by richvdh** [comment]
closing for lack of logs
